design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+using e.g. [`python-fuse` library](https://github.com/libfuse/python-fuse).
+It is not obvious that relying on FUSE is a good idea at all -- we need to conduct
+a study of pros/cons before committing to this path. There might be cases where
+FUSE would not be universally acceptable - there, we should offer the option to
+use an alternative, e.g. copy instead of stream the data.
+
+The interface should rely on a `DataSource` entity type. The role of the
+Provider is to configure the project and/or the user's environment in such a way
+as to make the `DataSource` available in a consistent, repeatable location with
+respect to the project and without boilerplate.
+
+### Local vs. hosted interactive sessions
+
+Relying on FUSE for providing filesystem access to external resources means that
+locally we do not require special privileges and can manage the data access for
+the user. In the docker-based remote sessions, the situation is more complicated
+because containers require elevated privileges to access FUSE on the host.
+Therefore we will need to expose a microservice that will provide management of
+external data sources for the user. The renku CLI would be aware that it is
+running in a hosted session and instead of executing a local command for
+mounting the data source, it would send off an API request to the service. The
+service itself could run in a sidecar or be integrated with datashim, which
+is currently used for providing S3 buckets in interactive sessions. If done via
+a sidecar, we could likely achive dynamic mounting of data sources, which would
+be fantastic for data exploration but comes with security trade-offs (sidecar needs
+to run with elevated privileges).
+
+Note that for NFS mounts, _root is still required_. A fuse-nfs project exists, but
+elevated privileges are needed for non-root users in order to use it.
+
+### Commands
+
+#### Adding external data to Renku Datasets
+
+The examples below are based on the existing `dataset` commands.
+
+```
+renku dataset add --create very-large-data s3://server/path/to/bucket
+-- prompt --
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+-- output --
+Credentials stored in ~/.renku/renku.ini
+```
+
+On the filesystem, the user would see `data/mydataset` as is normal for Renku
+datasets. Importantly, `data/mydataset` is also added to `.gitignore` to avoid
+exploding the git repository. 💥
+
+`s3://` is used here but we should also support at least `nfs://`. Renku should
+ask for credentials in a way that is appropriate for the type of resource that
+is being added.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show very-large-data
+-- output --
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+Data Sources:
+    - type: s3
+      url: s3://server/path/to/bucket
+      location: None
```

Comment:
Does this mean the data has not been staged anywhere? And `renku dataset stage ...` is required so that the data is actually available somewhere on the file system.

Reply:
that was my initial thinking. But maybe staging should be the default and you could use a flag to avoid.

Reply:
i've rephrased this bit slightly